src/c2pa/c2pa.py

```diff
@@ -1308,40 +1328,91 @@ def from_callback(
+        error_messages = {
+            'closed_error': "Signer is closed",
+            'cleanup_error': "Error during cleanup: {}",
+            'signer_cleanup': "Error cleaning up signer: {}",
+            'size_error': "Error getting reserve size: {}",
+            'callback_error': "Error in signer callback: {}",
+            'info_error': "Error creating signer from info: {}",
+            'invalid_data': "Invalid data for signing: {}",
+            'invalid_certs': "Invalid certificate data: {}",
+            'invalid_tsa': "Invalid TSA URL: {}",
+            'encoding_error': "Invalid UTF-8 characters in input: {}"
+        }
+
         # Validate inputs before creating
         if not certs:
             raise C2paError(
-                cls._error_messages['invalid_certs'].format("Missing certificate data"))
+                error_messages['invalid_certs'].format("Missing certificate data"))
 
         if tsa_url and not tsa_url.startswith(('http://', 'https://')):
             raise C2paError(
-                cls._error_messages['invalid_tsa'].format("Invalid TSA URL format"))
+                error_messages['invalid_tsa'].format("Invalid TSA URL format"))
 
         # Create a wrapper callback that handles errors and memory management
-        def wrapped_callback(data: bytes) -> bytes:
+        def wrapped_callback(
+                context,
+                data_ptr,
+                data_len,
+                signed_bytes_ptr,
+                signed_len):
+            # Returns 0 on error as this case is handled in the native code gracefully
+            # The reason is that otherwise we ping-pong errors between native code and Python code,
+            # which can become tedious in handling. So we let the native code deal with it and
+            # raise the errors accordingly, since it already checks the
+            # signature length for correctness.
             try:
+                if not data_ptr or data_len <= 0:
+                    # Error: invalid input, native code will handle if seeing
+                    # signature size being 0
+                    return 0
+
+                # Convert C pointer to Python bytes
+                data = bytes(data_ptr[:data_len])
                 if not data:
-                    raise ValueError("Empty data provided for signing")
-                return callback(data)
+                    # Error: empty data, native code will handle it!
+                    return 0
+
+                # Call the user's callback
+                signature = callback(data)
+                if not signature:
+                    # Error: empty signature, native code will handle that too!
+                    return 0
+
+                # Copy the signature back to the C buffer (since callback is
+                # sued in native code)
+                actual_len = min(len(signature), signed_len)
+                for i in range(actual_len):
```

Comment:
there's a spiffier way to do this, I think in the stream read code

Reply:
Done.